### PR TITLE
Stop the cheatcodes

### LINKED ISSUE #303: Add stop variants for `warp`,`roll`, `prank`

### PR DESCRIPTION
**Pull request by Radinyn**
Closes #303 

## Introduced changes

- Added `stop_roll`, `stop_warp` and `stop_prank`

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code

### CODE REVIEW

File: starknet-foundry/crates/cheatable-starknet/src/rpc.rs [hunk 1]
```diff
@@ -425,10 +425,7 @@ impl CheatableSyscallHandler<'_> {
             let ptr_cheated_exec_info = vm.add_memory_segment();
 
             // Initialize as old exec_info
-            let mut new_exec_info = vm
-                .get_continuous_range(execution_info_ptr, 5)
-                .unwrap()
-                .clone();
+            let mut new_exec_info = vm.get_continuous_range(execution_info_ptr, 5).unwrap();
```
Comment (Radinyn) [inline starknet-foundry/crates/cheatable-starknet/src/rpc.rs:428]: Clippy forced me to do this

**Review by piotmag769 — APPROVED** [review]

File: starknet-foundry/crates/forge/tests/integration/roll.rs [hunk 2]
```diff
@@ -169,3 +169,63 @@ fn start_roll_in_constructor_test() {
...
+            
+                start_roll(contract_address, 234);
+            
+                let block_number = dispatcher.get_block_number();
+                assert(block_number == 234, 'Wrong block number');
+
+                stop_roll(contract_address);
+
+                let block_number = dispatcher.get_block_number();
+                assert(block_number != 234, 'Block number  did not change back'');
```
Comment (piotmag769) [inline starknet-foundry/crates/forge/tests/integration/roll.rs:208]: I am wondering if we shouldn't get the original block number first, then roll, then stop roll and check if the number changed to the previous one . Same with other tests

**Review by piotmag769 — COMMENTED** [review]
Shouldn't we test stop_prank too?

File: starknet-foundry/crates/forge/tests/integration/roll.rs (hunk 2 above)
Comment (Arcticae) [inline starknet-foundry/crates/forge/tests/integration/roll.rs:208]: How does this work? The syntax seems wrong 😱 And the tests are passing? 🤔

Comment (piotmag769) [reply starknet-foundry/crates/forge/tests/integration/roll.rs:208]: Maybe they changed the way they are parsing? Hmmm

Comment (Arcticae) [reply starknet-foundry/crates/forge/tests/integration/roll.rs:208]: Should def be a syntax error

Comment (Radinyn) [reply starknet-foundry/crates/forge/tests/integration/roll.rs:208]: It's almost impossible to get such a block number by accident

Comment (Radinyn) [inline starknet-foundry/crates/forge/tests/integration/roll.rs:208]:
```suggestion
                assert(block_number != 234, 'Block number did not change back');
```

Comment (Radinyn) [reply starknet-foundry/crates/forge/tests/integration/roll.rs]: Fixed, not sure how does this work though, we should make an issue for the investigation of this topic.

Comment (piotmag769) [reply starknet-foundry/crates/forge/tests/integration/roll.rs]: https://github.com/foundry-rs/starknet-foundry/issues/338

File: crates/forge/tests/integration/prank.rs [hunk 3]
```diff
@@ -177,3 +177,67 @@ fn start_prank_in_constructor_test() {
...
+                let contract_address = deploy(prepared).unwrap();
+                let contract_address: ContractAddress = contract_address.try_into().unwrap();
+                let dispatcher = IPrankCheckerDispatcher { contract_address };
+            
+                let target_caller_address: felt252 = 123;
+                let target_caller_address: ContractAddress = target_caller_address.try_into().unwrap();
+
+                let old_caller_address = dispatcher.get_caller_address();
+
+                start_prank(target_caller_address, contract_address);
```
Comment (MaksymilianDemitraszek) [inline crates/forge/tests/integration/prank.rs:212]: Please add test cases for all of the cheatcodes:
- Calling on a not pranked contract
- Prank contract twice then stop

File: corelib/src/cheatcodes.cairo [hunk 4]
```diff
@@ -97,3 +97,18 @@ fn start_warp(contract_address: ContractAddress, block_timestamp: u64) {
     cheatcode::<'start_warp'>(array![contract_address_felt, block_timestamp_felt].span());
 }
 
+fn stop_roll(contract_address: ContractAddress) {
+    let contract_address_felt: felt252 = contract_address.into();
+    cheatcode::<'stop_roll'>(array![contract_address_felt].span());
+}
+
+fn stop_prank(target_contract_address: ContractAddress) {
```
Comment (MaksymilianDemitraszek) [inline corelib/src/cheatcodes.cairo:105]:
```suggestion
fn stop_prank(contract_address: ContractAddress) {
```

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]

File: corelib/src/cheatcodes.cairo (hunk 4 above)
Comment (Radinyn) [reply corelib/src/cheatcodes.cairo:105]: #336

File: crates/forge/tests/integration/prank.rs (hunk 3 above)
Comment (Radinyn) [reply crates/forge/tests/integration/prank.rs:212]: Calling on a not pranked contract is already tested afaik. I think I'd better to add double pranking in #336 instead

Comment (MaksymilianDemitraszek) [reply crates/forge/tests/integration/prank.rs]: > I think I'd better to add double pranking in https://github.com/foundry-rs/starknet-foundry/issues/336 instead

seems unrelated to the PR you mentioned, can you create an issue for it so we do not miss it?

**Review by MaksymilianDemitraszek — APPROVED** [review]

**Pull request merged**